### DISCUSSION TITLE
Separate world environment from post processing and add a better way to do post processing in 3D.

### DISCUSSION THREAD

**Discussion by tomankirilov** [godotengine/godot-proposals#8703] [3D]
I think world environment and post-processing should be separate nodes.
World Environment should deal with the sky, background light, and ambient light. Maybe with the fog too.
There should also exist a PostProcessing node. This node can deal with tone mapping, screen-space reflections,  screen-space AO, bloom etc. Making post-processing volumes is a good idea too, but I think just having an easy way to add a post-processing would already be a big step in a good direction.
On top of that I think that the way we are currently doing post-processing in 3D feels really hacky. Adding a plane, modifying the vertices to fit the camera in the vertex shader feels just wrong.
If we have a post-processing node it can probably contain a slot/slots for custom post-processing shaders?

**Comment by Calinou** [comment]
I think the current separation works well enough, but the issue is that there's no easy way to *globally* disable certain post-processing effects for performance reasons. However, there's a proposal that aims to address this: https://github.com/godotengine/godot-proposals/issues/5229

> Making post-processing volumes is a good idea too

This is discussed in https://github.com/godotengine/godot-proposals/issues/4466.

> On top of that I think that the way we are currently doing post-processing in 3D feels really hacky. Adding a plane, modifying the vertices to fit the camera in the vertex shader feels just wrong.

This is already being tracked in https://github.com/godotengine/godot-proposals/issues/2196.

**Reply by tomankirilov** [reply]
Hey, thanks again for the rapid-fire responses @Calinou ! Are you secretly part computer or something? Really appreciate you always being on the ball with my questions. Thanks for being awesome!
